### DISCUSSION TITLE
Online anomaly detection

### DISCUSSION THREAD

**Discussion by pranab** [online-ml/river#806] [Ideas]
Hi all,
I have a GitHub repo for anomaly detection with Spark/batch/scala  implementation for various algorithms for univariate, multivariate and time series data. I am proposing online incremental  reimplementation of them. It should be possible for most of the algorithms. It will be easy for all the stats based algorithms using running stat and some window. Your thoughts, comments are welcome. I already communicated with Max. He recommended posting in river/discussion. Here is the link to my repo. The project home page has links to blogs I wrote on various implementations

https://github.com/pranab/beymani

Pranab
https://www.linkedin.com/in/pkghosh/

**Comment by MaxHalford** [comment]
Hello @pranab! I suggest starting simple with a single method. Also, in general, we favor quality over quantity. Our goal isn't to implement every method out there, but instead to have those that generally considered to be battle-tested and popular. What would be a first model you might have in mind?

**Reply by pranab** [reply]
Since the algorithms depend on the type of data I am proposing 2 algorithms each for univariate, multivariate and time series data. For univariate  data, it will be zscore and robust zscore .I was planning to start with these 2

**Reply by MaxHalford** [reply]
Ok sounds good! I think we have to figure out some API changes, because the current API assumes each model accepts multivariate data, not univariate.

**Reply by pranab** [reply]
My understanding is that feature input is a dictionary. So it should work for 1 feature (univariate)  as well

**Reply by pranab** [reply]
Running stat (mean and variance) is required for zscore based anomaly detection. In Mean Var class constructors the current stat values are set to zero. I need to make a change so that stat values from the previous run can be passed to the constructor as arguments for the next run. update() will use them.

**Reply by smastelini** [reply]
Hi @pranab. You can rely on the `_from_state` class method to achieve what you want:

https://github.com/online-ml/river/blob/3732f700da72642afe54095d4b252b05c5018c7d/river/stats/var.py#L67-L73

I also had to do a similar action in the past and relied on them :)

**Reply by pranab** [reply]
Hi @smastelini thanks. RollingVar and RollingMean don't have those methods. I will add them

**Reply by pranab** [reply]
Hi @MaxHalford I created a pull request for zscore based anomaly detection

**Comment by MaxHalford** [comment]
Cool! I'm on holiday right until the 26th, I'll check it out soon :)

**Reply by pranab** [reply]
@MaxHalford  not sure what do with the 2 errors under Action from my pull request

**Reply by MaxHalford** [reply]
Answered!
